#18 - Systems with outputs

mforets opened the pull request:
Closes #18.

mforets commented:
i've hesitated about `...maps` because one can argue that `...DiscreteSystem` can be used for that purpose. perhaps we can make the `MT` in the output be optional, eg.

```julia
struct SystemWithOutput{ST<:AbstractSystem, MT}   # no restriction on MT
    system::ST
    outputmap::MT
end
```

schillic commented on src/maps.jl:
```diff
@@ -0,0 +1,91 @@
+end
+
+"""
+    AffineMap
```
```suggestion
    AffineControlMap
```

schillic commented:
I do not understand what you want to say, sorry.

blegat commented on src/maps.jl:
```diff
@@ -0,0 +1,91 @@
+"""
+    AffineMap
+
+An affine map
```
affine control

blegat commented on src/maps.jl:
```diff
@@ -0,0 +1,91 @@
+end
+
+"""
+    SystemWithOutput
```
Maybe add parameters here and add in the `Fields` section "of type `ST`"

blegat commented on src/maps.jl:
```diff
@@ -0,0 +1,91 @@
+"""
+    LinearControlMap
+
+An affine map
```
affine -> linear control

schillic commented:
>```julia
>struct LinearMap{T, MT<:AbstractMatrix} <: AbstractMap
>    A::MT
>end
>```

You need to use `T` here :wink:

blegat commented:
Yes it should be `AbstractMatrix{T}`

mforets commented:
i've added 2 more types `ConstrainedLinearControlMap` and `ConstrainedAffineControlMap` making `LinearControlMap` and `AffineControlMap` to be unconstrained (as it is the case for systems).

then i've added unit tests and docs.

new methods `outputdim` and `outputmap` have been defined for a `SystemWithOutput`. 

an LTI system can be constructed as a `LTISystem(A, B, C, D)` or a `LTISystem(A, B, C, D, X, U)` if one wants to specify both state and input constraints.

ready from my side.

mforets commented on src/MathematicalSystems.jl:
```diff
@@ -14,7 +14,8 @@ export AbstractSystem,
        inputdim,
-       inputset
+       inputset,
+       outputdim
```
ok

blegat commented on src/outputs.jl:
```diff
@@ -0,0 +1,87 @@
+"""
+function LinearTimeInvariantSystem(A, B, C, D)
+    system = LinearControlContinuousSystem(A, B)
+    outputmap = LinearControlMap(A, B)
```
It should be `C`, `D` here

mforets replied:
oops

blegat replied:
We should add a test to catch that

schillic commented on test/outputs.jl:
```diff
@@ -2,21 +2,25 @@
     s = LTISystem(A, B, C, D)
+
+    # check the LTI system's outputmap fields 
+    @test outputmap(s).A == C
```
You could add the analogous tests for `s.system`.
